cli/cmd/upgrade.go

```diff
@@ -0,0 +1,207 @@
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/golang/protobuf/jsonpb"
+	pb "github.com/linkerd/linkerd2/controller/gen/config"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/linkerd/linkerd2/pkg/tls"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type (
+	upgradeOptions struct{ *installOptions }
+)
+
+func newUpgradeOptionsWithDefaults() *upgradeOptions {
+	return &upgradeOptions{newInstallOptionsWithDefaults()}
+}
+
+func newCmdUpgrade() *cobra.Command {
+	options := newUpgradeOptionsWithDefaults()
+
+	cmd := &cobra.Command{
+		Use:   "upgrade [flags]",
+		Short: "Output Kubernetes configs to upgrade an existing Linkerd control plane",
+		Long:  "Output Kubernetes configs to upgrade an existing Linkerd control plane.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			k, err := newKubernetes()
+			if err != nil {
+				return err
+			}
+
+			values, configs, err := options.buildFromCluster(k)
+			if err != nil {
+				return err
+			}
+
+			return values.render(os.Stdout, configs)
+		},
+	}
+
+	options.configure(cmd)
+	return cmd
+}
+
+func newKubernetes() (*kubernetes.Clientset, error) {
+	api, err := k8s.NewAPI(kubeconfigPath, kubeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	return kubernetes.NewForConfig(api.Config)
+}
+
+// fetchInstallValuesFromCluster checks the kubernetes API to fetch an existing
+// linkerd configuration.
+//
+// This bypasses the public API so that we can access secrets and validate permissions.
+func (options *upgradeOptions) buildFromCluster(k *kubernetes.Clientset) (*installValues, *pb.All, error) {
+	if options.ignoreCluster {
+		return nil, nil, errors.New("--ignore-cluster cannot be used with upgrade")
+	}
+
+	controllerReplicas := uint(1)
+	identityReplicas := uint(1)
+	selector := fmt.Sprintf("%s in (controller, identity)", k8s.ControllerComponentLabel)
+	deploys, err := k.ExtensionsV1beta1().Deployments(controlPlaneNamespace).
+		List(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, deploy := range deploys.Items {
+		if deploy.Spec.Replicas == nil {
+			continue
+		}
+		r := uint(*deploy.Spec.Replicas)
+
+		switch deploy.GetLabels()[k8s.ControllerComponentLabel] {
+		case "controller":
+			controllerReplicas = r
+		case "identity":
+			identityReplicas = r
+		}
+	}
+
+	// Fetch the existing cluster configs or exit.
+	configs := &pb.All{Global: &pb.Global{}, Proxy: &pb.Proxy{}}
+	configMap, err := k.CoreV1().ConfigMaps(controlPlaneNamespace).
+		Get(k8s.ConfigConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	uj := jsonpb.Unmarshaler{}
+	if err := uj.Unmarshal(strings.NewReader(configMap.Data["global"]), configs.Global); err != nil {
```

Comment:
would like to see these "global" and "proxy" strings as constants and templatized in `config.yaml`.